br/pkg/lightning/backend/external/iter.go

```diff
@@ -205,20 +221,49 @@ func (i *mergeIter[T, R]) next() bool {
 		i.hotspotMap[i.lastReaderIdx] = i.hotspotMap[i.lastReaderIdx] + 1
 		i.checkHotspotCnt++
 
-		checkPeriod := 1000
-		// check hot point every checkPeriod times
-		if i.checkHotspotCnt == checkPeriod {
+		// check hotspot every checkPeriod times
+		if i.checkHotspotCnt == checkHotspotPeriod {
+			oldHotspotIdx := i.lastHotspotIdx
+			i.lastHotspotIdx = -1
 			for idx, cnt := range i.hotspotMap {
-				(*i.readers[idx]).setReadMode(cnt > (checkPeriod / 2))
+				// currently only one reader will become hotspot
+				if cnt > (checkHotspotPeriod / 2) {
+					i.lastHotspotIdx = idx
+					break
+				}
+			}
+			// we are going to switch concurrent reader and free its memory. Clone
+			// the fields to avoid use-after-free.
+			if oldHotspotIdx != i.lastHotspotIdx {
+				if i.elemFromHotspot != nil {
+					(*i.elemFromHotspot).cloneInnerFields()
+					i.elemFromHotspot = nil
+				}
+			}
+
+			for idx, rp := range i.readers {
+				if rp == nil {
+					continue
+				}
+				isHotspot := i.lastHotspotIdx == idx
+				err := (*rp).switchConcurrentMode(isHotspot)
+				if err != nil {
+					i.err = err
+					return false
+				}
```

Comment:
We only need to swtich at most two readers. Is there a plan to support concurrent read on multiple readers?

Reply:
Yes I hope so. After we log the dropped data we will can better evaluate current strategy, and see if it's needed to learn more about external merge sort prefetch